asio: 1.12 -> 1.16

siriobalmelli opened the pull request:
1.16 will now build on Darwin, properly mark it as such.

Keep asio_1_10 and asio_1_12 for applications which cannot easily upgrade.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

###### Motivation for this change
Another package dependency on `asio` broken in Darwin (will not compile).
The latest version will compile: upgraded.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions (Ubuntu 18.04)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev 931cd746"`
- [n/a] Tested execution of all binary files (usually in `./result/bin/`)
- [n/a] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

siriobalmelli committed:
asio: 1.12 -> 1.16

1.16 will now build on Darwin, properly mark it as such.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>
Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

jonringer commented on pkgs/development/libraries/asio/generic.nix:
```diff
@@ -21,7 +21,7 @@ stdenv.mkDerivation {
     description = "Cross-platform C++ library for network and low-level I/O programming";
     license = licenses.boost;
-    broken = stdenv.isDarwin;  # test when updating to >=1.12.1
+    broken = stdenv.isDarwin && version < "1.16.1";
```
```suggestion
    broken = stdenv.isDarwin && stdenv.lib.versionOlder version "1.16.1";
```
otherwise it's doing a simple string comparison:
```
nix-repl> lib.versionOlder "1.2" "1.11"
true
nix-repl> "1.2" < "1.11"
false
```

siriobalmelli replied:
Ooh, thank you!
I learned something with that

jonringer commented:
Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the commits.

This can be done with:
```
git reset HEAD~1
git add -- pkgs/
git commit --amend --no-edit
git push ... ... --force
```

veprbl commented:
Commit description should read: "asio: 1.12 -> 1.16". The `asio_12` is not referenced within the nixpkgs, the policy is to delete such packages.

siriobalmelli commented:
> Commit description should read: "asio: 1.12 -> 1.16". The `asio_12` is not referenced within the nixpkgs, the policy is to delete such packages.

Thank you, review much appreciated.

- rebased on top of latest master
- removed asio 1.12
- edited commit message

veprbl commented:
Still missing a colon. We can change this in squash merge, but that will break your gpg signature.

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -11285,9 +11285,9 @@ in
+  asio = asio_1_16;
   asio_1_10 = callPackage ../development/libraries/asio/1.10.nix { };
-  asio_1_12 = callPackage ../development/libraries/asio/1.12.nix { };
+  asio_1_16 = callPackage ../development/libraries/asio/1.16.nix { };
```
Might as well drop the `asio_1_16` attribute. Out-of-tree expressions can always check asio.version if they need to.
```suggestion
  asio = callPackage ../development/libraries/asio/1.16.nix { };
```

siriobalmelli replied:
Thank you :)

No worries, will squash commits and add colon.
Stand by for force-push.

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -11285,9 +11285,8 @@ in
-  asio = asio_1_12;
   asio_1_10 = callPackage ../development/libraries/asio/1.10.nix { };
-  asio_1_12 = callPackage ../development/libraries/asio/1.12.nix { };
+  asio = callPackage ../development/libraries/asio/1.16.nix { };
```
Sorry for not requesting this earlier. Can we also do a file rename please.
```suggestion
  asio = callPackage ../development/libraries/asio/default.nix { };
```

veprbl commented:
Thank you!
